src/hooks/useMultiplayerState.ts

```diff
@@ -63,14 +66,98 @@ export function useMultiplayerState(roomId: string) {
 		[],
 	);
 
-	const onMount = useCallback(
-		(app: TldrawApp) => {
-			app.loadRoom(roomId);
-			app.pause();
-			setAppInstance(app);
-		},
-		[roomId],
-	);
+	async function openFromFileSystem(): Promise<null | {
+		fileHandle: FileSystemFileHandle | null;
+		document: TDDocument;
+	}> {
+		const blob = await fileOpen({
+			description: 'Tldraw File',
+			extensions: [`.tldr`],
+			multiple: false,
+		});
+
+		if (!blob) return null;
+
+		const json: string = await new Promise((resolve) => {
+			const reader = new FileReader();
+			reader.onloadend = () => {
+				if (reader.readyState === FileReader.DONE) {
+					resolve(reader.result as string);
+				}
+			};
+			reader.readAsText(blob, 'utf8');
+		});
+
+		const file: TDFile = JSON.parse(json);
+
+		const fileHandle = blob.handle ?? null;
+
+		return {
+			fileHandle,
+			document: file.document,
+		};
+	}
+
+	const updateYMapsInTransaction = (
+		shapes: Record<string, TDShape | undefined>,
+		bindings: Record<string, TDBinding | undefined>,
+		assets: Record<string, TDAsset | undefined>,
+	) => {
+		doc.transact(() => {
+			Object.entries(shapes).forEach(([id, shape]) => {
+				if (!shape) {
+					yShapes.delete(id);
+				} else {
+					yShapes.set(shape.id, shape);
+				}
+			});
+			Object.entries(bindings).forEach(([id, binding]) => {
+				if (!binding) {
+					yBindings.delete(id);
+				} else {
+					yBindings.set(binding.id, binding);
+				}
+			});
+			Object.entries(assets).forEach(([id, asset]) => {
+				if (!asset) {
+					yAssets.delete(id);
+				} else {
+					yAssets.set(asset.id, asset);
+				}
+			});
+		});
+	};
+
+	const onMount = useCallback((app: TldrawApp) => {
+		app.loadRoom(roomId);
+		app.pause();
+		setAppInstance(app);
+
+		app.openProject = async () => {
+			try {
+				const result = await openFromFileSystem();
+				if (!result) {
+					throw Error();
+				}
+
+				const { document } = result;
+
+				yShapes.clear();
+				yBindings.clear();
+				yAssets.clear();
+
+				updateYMapsInTransaction(
+					document.pages.page.shapes,
+					document.pages.page.bindings,
+					document.assets,
+				);
+
+				app.zoomToFit();
+			} catch (e) {
+				console.error(e);
```

Comment:
In generell:
Using console must avoid. Please introduce a logger framework/package for it. This is the sync write operations that shorty blog the node process. By high traffic it can be reduce the response time a lot.

In your case:
But your code is only executed in brower correct?

Reply:
Okay, i can add a logger package, but which one?
Loglevel? Winston?